[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Boyu-Zhang-UOI/pml-f2026-notebooks/blob/main/08-llms/llms_in_practice.ipynb)

# LLMs in Practice: Prompting, RAG & LoRA

**CS 4771/5771 · Python for Machine Learning · Module 7: Generative AI & LLMs**

This notebook goes with Session 28 (*LLMs in Practice*), an optional session taught
only if class time remains; either way it is after-class material. It runs on the
**free Google Colab tier**: a T4 GPU helps but is not required, because every model
here is small.

It uses the slides' running example, the documentation of a research computing
cluster, so the retrieval numbers below are the ones on the slides.

We will:

1. Prompt a small open-weight LLM (zero-shot vs. few-shot)
2. Build a **mini-RAG** system on the cluster documentation: TF-IDF retrieval, an
   abstention threshold, an embedding model, and prompt assembly
3. Attach **LoRA** adapters with the `peft` library, count what they train, and run
   a few training steps
4. Close with a checklist for verifying AI-generated output

## 0. Setup

Uncomment and run the cell below once per Colab session. (On Colab, `transformers`
and `torch` are usually preinstalled; `sentence-transformers` and `peft` are not.)
On your own machine, `requirements-nlp-advanced.txt` installs all of them.

In [1]:
# %pip install -q transformers sentence-transformers datasets peft accelerate

import torch
import numpy as np
import transformers

# Hide the library's version chatter (deprecation notices repeated on every
# call); errors still show.
transformers.logging.set_verbosity_error()

# CUDA when Colab gives you a GPU, CPU otherwise. Apple's "mps" backend is
# deliberately not used here: this model's grouped-query attention hits a Metal
# kernel bug that kills the process rather than raising.
device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Running on: {device}")

Running on: cpu


## 1. Prompting a small open model

We use **Qwen2.5-0.5B-Instruct**, a 0.5B-parameter instruction-tuned model that
downloads in about a gigabyte and runs even on the Colab CPU (slowly) or a T4 GPU
(comfortably). The model is fetched by name, without a pinned revision, so if the
upstream files change your replies may differ from the saved ones below.

> **Substitutions:** any small instruct model works the same way, e.g.
> `HuggingFaceTB/SmolLM2-360M-Instruct` or `TinyLlama/TinyLlama-1.1B-Chat-v1.0`.
> Expect small models to make mistakes: that is part of the lesson.

In [2]:
from transformers import pipeline

MODEL_NAME = "Qwen/Qwen2.5-0.5B-Instruct"

generator = pipeline(
    "text-generation",
    model=MODEL_NAME,
    device=device,              # the device chosen above, not a guess
    dtype="auto",
)


def chat(prompt, max_new_tokens=80):
    """Send a single-turn chat prompt and return the assistant's reply."""
    messages = [{"role": "user", "content": prompt}]
    out = generator(messages, max_new_tokens=max_new_tokens, do_sample=False)
    return out[0]["generated_text"][-1]["content"].strip()


print(chat("In one sentence, what does a diffusion model learn to do?"))

A diffusion model learns to predict the likelihood of new information spreading through social networks or other interconnected systems over time.


Read that reply against Session 27 before moving on. A diffusion model learns to
undo a noising process: a network predicts the noise added to an image, and a
reverse loop walks from noise to a sample. In the saved run the model answered
about information spreading through social networks instead: fluent, confident,
and about the wrong kind of diffusion. Nothing in the reply signals the mistake.

### Zero-shot vs. few-shot

The task: classify a course-feedback comment into exactly one of three labels.
First we just *describe* the task (zero-shot); then we *show* the pattern with a few
worked examples (few-shot, also called in-context learning).

In [3]:
comment = "The homework took forever, but I finally understand backpropagation."

zero_shot = f"""Classify the following course feedback as POSITIVE, NEGATIVE, or MIXED.
Reply with exactly one word.

Feedback: {comment}"""

print("ZERO-SHOT:", chat(zero_shot, max_new_tokens=10))

ZERO-SHOT: POSITIVE


In [4]:
few_shot = f"""Classify course feedback as POSITIVE, NEGATIVE, or MIXED.
Reply with exactly one word.

Feedback: The lectures were clear and the pace was perfect.
Label: POSITIVE

Feedback: I could never hear the instructor and the slides were outdated.
Label: NEGATIVE

Feedback: Great projects, though the grading felt slow.
Label: MIXED

Feedback: {comment}
Label:"""

print("FEW-SHOT:", chat(few_shot, max_new_tokens=10))

FEW-SHOT: POSITIVE


In the saved run both prompts answer POSITIVE, although a complaint followed by a
win reads as MIXED. The examples changed nothing here: the one-word format was
already right, and the judgment did not move. Few-shot examples shape *how* a model
answers; they do not guarantee *what* it concludes.

**Try it:** change `comment` to something sarcastic ("Oh great, *another* proctored
exam") and compare the two prompts again. Two more techniques to experiment with:

- **Chain-of-thought:** append "Think step by step before giving the label." (allow
  more `max_new_tokens`). Read the steps as a place to look for errors, not as the
  model's actual reasoning.
- **Structured output:** ask for `{"label": ..., "confidence": ...}` as JSON and parse
  it with `json.loads`, then validate the fields as the slides' `parse_record` does,
  and count how often a small model breaks the format.

## 2. Mini-RAG on the cluster documentation

The model above has never seen *your* documents. Ours are the slides' running
example: eight paragraphs (326 words) of documentation for a research computing
cluster. The cluster is made up; its numbers (12 hours on the standard partition,
4 GB per core, 30 days on scratch) are the ones the slides use.

RAG retrieves the relevant chunks at question time and puts them in the prompt.
Each paragraph is one chunk here, the chunking the slides found best on this
corpus. The cell also lists eight questions the documents answer and six they do
not.

In [5]:
chunks = [
    ('Cluster accounts are requested through the research computing '
     'portal. Every request must name a faculty sponsor who approves it '
     'before the account is opened. Accounts are reviewed once a year and '
     'lapse if the sponsor does not renew them.'),
    ('Every user receives a home directory on the shared filesystem. The '
     'home directory quota is 50 GB and it is not intended for datasets. '
     'Large inputs belong in the project space, which is allocated per '
     'group.'),
    ('The scheduler assigns processor cores and memory to each job. The '
     'default memory allocation is 4 GB per core, and a job that exceeds '
     'its request is killed rather than throttled. Ask for more with the '
     'memory flag when the job loads a large table into RAM.'),
    ('Work is submitted to one of three partitions. The standard partition '
     'has a wall time limit of 12 hours, the long partition allows five '
     'days, and the debug partition allows 30 minutes with a higher '
     'scheduling priority. Choose the shortest partition that fits, '
     'because shorter jobs start sooner.'),
    ('Accelerated nodes carry graphics processors. A single job may '
     'request at most 4 GPUs, and requests larger than that are rejected '
     'at submission. GPU jobs are billed against the group allocation at '
     'four times the processor rate.'),
    ('Software is provided through environment modules. Python '
     'environments are managed with conda, and a module load is required '
     'before the interpreter is on the path. Compiled libraries are built '
     'against the system compiler, so mixing toolchains in one environment '
     'is unsupported.'),
    ('Running work can be inspected from any login node. Use the queue '
     'command to list your jobs and scancel with the job identifier to '
     'stop one. A job that is stopped this way still counts the processor '
     'time it has already used.'),
    ('Two storage tiers have different retention rules. Scratch files are '
     'deleted after 30 days without access, and nothing on scratch is '
     'backed up. Project space is snapshotted nightly and the snapshots '
     'are kept for two weeks.'),
]

answerable = [
    ('Who has to approve a new cluster account?', 'faculty sponsor'),
    ('What is the home directory quota?', '50 GB'),
    ('How much memory is allocated per core by default?', '4 GB per core'),
    ('What is the wall time limit on the standard partition?', '12 hours'),
    ('How many GPUs can a single job request?', '4 GPUs'),
    ('How are Python environments managed?', 'conda'),
    ('Which command stops a running job?', 'scancel'),
    ('How long are scratch files kept?', '30 days'),
]

out_of_scope = [
    'What does an hour of cloud GPU time cost?',
    'How do I reset my university email password?',
    'Where do I buy a parking permit?',
    'Which building houses the machine shop?',
    'How do I reserve a conference room?',
    'When does the campus shuttle run?',
]

print(f"{len(chunks)} chunks, {sum(len(c.split()) for c in chunks)} words; "
      f"{len(answerable)} answerable and {len(out_of_scope)} out-of-scope questions")

8 chunks, 326 words; 8 answerable and 6 out-of-scope questions


### First, the retriever that needs no download

**TF-IDF** (term frequency, inverse document frequency) gives each word a weight
that grows with its count in a chunk and shrinks the more chunks use it.
**Cosine similarity** compares the directions of two such vectors: the dot product
after dividing each by its length. This is the code on the slide *Retrieval in
NumPy*, so the scores match it. (`TfidfVectorizer` already returns rows of length
1; the divisions keep the code right for vectors that are not normalized.)

In [6]:
from sklearn.feature_extraction.text import TfidfVectorizer

question = "How much memory is allocated per core by default?"
vec = TfidfVectorizer(stop_words="english")
doc_mat = vec.fit_transform(chunks).toarray()
print("document matrix:", doc_mat.shape)
q_vec = vec.transform([question]).toarray()
docs = doc_mat / np.linalg.norm(doc_mat, axis=1, keepdims=True)
q = q_vec / np.linalg.norm(q_vec)
scores = (docs @ q.T).ravel()        # cosine, one per chunk
for i in np.argsort(-scores)[:3]:
    print(f"{scores[i]:.3f}  chunk {i}: {chunks[i][:30]}...")

document matrix: (8, 143)
0.458  chunk 2: The scheduler assigns processo...
0.105  chunk 1: Every user receives a home dir...
0.000  chunk 0: Cluster accounts are requested...


### An abstention threshold, chosen from measured scores

A retriever always returns its best chunk, even when nothing is relevant. So the
system needs a rule for when the best is not good enough. The slides set a cutoff
of 0.25 from the measured top-1 scores of both question groups; here are those
measurements.

In [7]:
def tfidf_scores(question):
    """Cosine similarity of one question to every chunk (TF-IDF rows are unit length)."""
    return doc_mat @ vec.transform([question]).toarray().ravel()


THRESHOLD = 0.25

for label, questions in [("in corpus", [q for q, _ in answerable]),
                         ("out of scope", out_of_scope)]:
    print(label)
    for question in questions:
        best = tfidf_scores(question).max()
        verdict = "answer" if best >= THRESHOLD else "ABSTAIN"
        print(f"  {best:.3f}  {verdict:<7}  {question}")

in corpus
  0.298  answer   Who has to approve a new cluster account?
  0.608  answer   What is the home directory quota?
  0.458  answer   How much memory is allocated per core by default?
  0.546  answer   What is the wall time limit on the standard partition?
  0.406  answer   How many GPUs can a single job request?
  0.361  answer   How are Python environments managed?
  0.434  answer   Which command stops a running job?
  0.438  answer   How long are scratch files kept?
out of scope
  0.141  ABSTAIN  What does an hour of cloud GPU time cost?
  0.000  ABSTAIN  How do I reset my university email password?
  0.000  ABSTAIN  Where do I buy a parking permit?
  0.000  ABSTAIN  Which building houses the machine shop?
  0.000  ABSTAIN  How do I reserve a conference room?
  0.211  ABSTAIN  When does the campus shuttle run?


Every in-corpus question scores at least 0.298 and every out-of-scope one at most
0.211, so 0.25 answers 8 of 8 and abstains on 6 of 6.

**Try it (after class):** set `THRESHOLD` to 0.15, then to 0.35, and note which
questions change sides. The cutoff is a measurement on *this* corpus with *this*
retriever, not a constant.

### A trained embedding model

TF-IDF matches spelling. A sentence-embedding model maps text to a dense vector
where meaning decides distance. `all-MiniLM-L6-v2` (about 90 MB) gives 384 numbers
per text, and the same NumPy arithmetic scores them. The question below shares no
word with the documents.

In [8]:
from sentence_transformers import SentenceTransformer

embedder = SentenceTransformer("all-MiniLM-L6-v2")
chunk_emb = embedder.encode(chunks, normalize_embeddings=True)
print("embedding matrix:", chunk_emb.shape)


def embedding_scores(question):
    """Cosine similarity of one question to every chunk (normalized embeddings)."""
    return chunk_emb @ embedder.encode([question], normalize_embeddings=True)[0]


paraphrase = "What's the longest a task may run?"
tf = tfidf_scores(paraphrase)
em = embedding_scores(paraphrase)
print(f"\n{paraphrase}")
print(f"  TF-IDF best score over all chunks: {tf.max():.3f}")
order = np.argsort(-em)
print(f"  embedding: chunk {order[0]} ({em[order[0]]:.3f}) "
      f"{chunks[order[0]][:45]}...")
print(f"             runner-up chunk {order[1]} ({em[order[1]]:.3f})")

embedding matrix: (8, 384)

What's the longest a task may run?
  TF-IDF best score over all chunks: 0.000
  embedding: chunk 3 (0.532) Work is submitted to one of three partitions....
             runner-up chunk 6 (0.391)


TF-IDF scores every chunk 0 because no word overlaps; the embedding model puts the
partitions paragraph first. Now the catch: the threshold belongs to the
retriever. Here are the same fourteen questions scored with embeddings.

In [9]:
inside = {q: embedding_scores(q).max() for q, _ in answerable}
outside = {q: embedding_scores(q).max() for q in out_of_scope}
low_in = min(inside, key=inside.get)
high_out = max(outside, key=outside.get)
print(f"in corpus:    {min(inside.values()):.3f} to {max(inside.values()):.3f}  "
      f"(lowest: {low_in})")
print(f"out of scope: {min(outside.values()):.3f} to {max(outside.values()):.3f}  "
      f"(highest: {high_out})")
passed = [q for q, s in outside.items() if s >= THRESHOLD]
print(f"out-of-scope questions TF-IDF's cutoff of {THRESHOLD} would let through: {len(passed)}")

in corpus:    0.453 to 0.806  (lowest: Which command stops a running job?)
out of scope: 0.147 to 0.493  (highest: What does an hour of cloud GPU time cost?)
out-of-scope questions TF-IDF's cutoff of 0.25 would let through: 2


With embeddings the two groups overlap: the question about the price of cloud GPU
time, which the documents cannot answer, outscores an in-corpus question, so no
cutoff separates them cleanly, and TF-IDF's 0.25 would let two out-of-scope
questions through. A new retriever means new measurements and a new threshold.

### Assembling the prompt, and abstaining before the model is called

`build_prompt` is the function on the slide *Assembling the Prompt*: numbered
chunks the answer can cite, a delimiter around the untrusted document text, and an
exact refusal string. `rag_answer` adds the first layer of the grounding defenses:
below the threshold it never calls the model at all.

In [10]:
def build_prompt(question, chunks):
    """Instruction outside, documents inside."""
    ctx = "\n\n".join(f"[{i}] {c}"
                      for i, c in enumerate(chunks, 1))
    return (
        "Answer the question using only the numbered\n"
        "context below. Cite the numbers you used, as [2].\n"
        "If the context does not contain the answer, reply\n"
        "exactly: not covered by the documents.\n\n"
        "<context>\n" + ctx + "\n</context>\n\n"
        f"Question: {question}"
    )


def rag_answer(question, k=2):
    scores = tfidf_scores(question)
    if scores.max() < THRESHOLD:
        return f"(abstained: best score {scores.max():.3f} < {THRESHOLD})"
    top = [chunks[i] for i in np.argsort(-scores)[:k]]
    return chat(build_prompt(question, top), max_new_tokens=60)


question = "How much memory is allocated per core by default?"
print("NO CONTEXT:", chat(question, max_new_tokens=60), "\n")
for k in (2, 1):
    print(f"WITH RAG, top {k}:", rag_answer(question, k=k))
print("\nOUT OF SCOPE:", rag_answer("Where do I buy a parking permit?"))

NO CONTEXT: The amount of memory (RAM) that can be allocated per core by default for an ECS instance depends on the specific version and configuration of the Elastic Compute Service (ECS) service you are using.

For example:

- **Ubuntu 18.04** or later versions: The default allocation 



WITH RAG, top 2: [2]


WITH RAG, top 1: The default memory allocation is 4 GB per core.

OUT OF SCOPE: (abstained: best score 0.000 < 0.25)


**Things to notice** (in the saved run; a larger model, or a GPU, may reply
differently):

- Without the documents the model still answers, fluently, about some cloud
  provider's machines: a confident guess with nothing behind it.
- With the top two chunks, the prompt on the slide *Assembling the Prompt*, the 0.5B
  model replied with nothing but a citation, and the wrong one: `[1]` is the
  scheduler paragraph that holds "4 GB per core", `[2]` the home-directory one. With
  only the top chunk it answered correctly but cited nothing. This is why the
  grounding defenses check citations in code, and why retrieval and generation are
  evaluated separately: here retrieval was right and generation was not.
- The out-of-scope question never reached the model. The threshold, not the model's
  good behavior, decided to abstain.

**Try it:** write the citation check from the slides: pull the `[n]` markers out of a
reply with a regular expression and confirm that chunk `n` contains the answer.
- The retriever and the generator are separate, swappable parts. A vector database
  (FAISS, Chroma, pgvector) replaces the NumPy search at a million chunks; the
  arithmetic is the same.

## 3. Parameter-efficient fine-tuning with LoRA

RAG supplies *knowledge*; fine-tuning changes *behavior*. LoRA freezes each adapted
weight matrix $W$ and learns a low-rank update $\Delta W = BA$, with $A$ of shape
$r \times d_{\text{in}}$ and $B$ of shape $d_{\text{out}} \times r$. Below, adapters go on
the query and value projections of the same Qwen model, with the configuration on
the slide *LoRA in Code, and QLoRA*. The weights come from the cache, so nothing new
is downloaded.

In [11]:
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import LoraConfig, get_peft_model

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
base_model = AutoModelForCausalLM.from_pretrained(MODEL_NAME, dtype=torch.float32)

lora_config = LoraConfig(
    r=8,                        # rank of B @ A
    lora_alpha=32,              # B @ A is scaled by lora_alpha / r = 4
    target_modules=["q_proj", "v_proj"],
    lora_dropout=0.05,
    task_type="CAUSAL_LM",
)
model = get_peft_model(base_model, lora_config)
model.print_trainable_parameters()

# Check the count by hand: r * (d_in + d_out) per adapted matrix. Qwen2.5-0.5B has
# 24 layers and hidden size 896; with two key/value heads of 64, v_proj outputs 128.
cfg = base_model.config
d, kv = cfg.hidden_size, cfg.num_key_value_heads * (cfg.hidden_size // cfg.num_attention_heads)
by_hand = cfg.num_hidden_layers * (8 * (d + d) + 8 * (d + kv))
print(f"by hand: {cfg.num_hidden_layers} layers x (q {8 * (d + d):,} + v {8 * (d + kv):,}) = {by_hand:,}")

trainable params: 540,672 || all params: 494,573,440 || trainable%: 0.1093
by hand: 24 layers x (q 14,336 + v 8,192) = 540,672


`print_trainable_parameters` counts the adapters in its total, which is why the
total is larger than the base model's. The slides count a 7B model the same way.
That model cannot be downloaded here, but its *shapes* can be built on PyTorch's
`meta` device, which allocates no memory, and peft counts them just the same.

In [12]:
from transformers import LlamaConfig

llama2_7b = LlamaConfig(vocab_size=32000, hidden_size=4096, intermediate_size=11008,
                        num_hidden_layers=32, num_attention_heads=32,
                        num_key_value_heads=32, tie_word_embeddings=False)
with torch.device("meta"):
    big = AutoModelForCausalLM.from_config(llama2_7b)
print(f"base model: {sum(p.numel() for p in big.parameters()):,} parameters")
# A fresh config: the one above is now bound to the Qwen model.
big_config = LoraConfig(r=8, lora_alpha=32, target_modules=["q_proj", "v_proj"],
                        lora_dropout=0.05, task_type="CAUSAL_LM")
get_peft_model(big, big_config).print_trainable_parameters()

base model: 6,738,415,616 parameters
trainable params: 4,194,304 || all params: 6,742,609,920 || trainable%: 0.0622


### Optional: a few training steps

This cell teaches the model a house style from six examples: far too little data to
matter, but it runs the whole LoRA training loop in about a minute, even on a CPU.
Real runs use `transformers.Trainer` (or `trl`'s `SFTTrainer`) with a proper dataset;
either way only the adapter weights receive gradients.

**This cell is optional: skip it if you are short on time.**

In [13]:
# --- OPTIONAL TRAINING CELL ---
import os
import tempfile

tiny_dataset = [
    "Q: What is overfitting? A: In ML terms: memorizing noise instead of learning signal.",
    "Q: What is a tensor? A: In ML terms: an n-dimensional array that autograd can track.",
    "Q: What is a loss function? A: In ML terms: a score for how wrong the model is.",
    "Q: What is gradient descent? A: In ML terms: repeated steps against the gradient.",
    "Q: What is regularization? A: In ML terms: a penalty that limits model complexity.",
    "Q: What is an embedding? A: In ML terms: a learned vector that encodes meaning.",
]

batch = tokenizer(tiny_dataset, return_tensors="pt", padding=True)
labels = batch["input_ids"].clone()
labels[batch["attention_mask"] == 0] = -100    # no loss on padding positions
batch["labels"] = labels

torch.manual_seed(0)
model.train()
optimizer = torch.optim.AdamW(
    (p for p in model.parameters() if p.requires_grad), lr=2e-4
)

for step in range(10):                       # a real run would take thousands of steps
    outputs = model(**batch)
    outputs.loss.backward()
    optimizer.step()
    optimizer.zero_grad()
    print(f"step {step:2d}  loss = {outputs.loss.item():.3f}")

# The trained adapter is saved on its own, separately from the base model.
with tempfile.TemporaryDirectory() as tmp:
    model.save_pretrained(tmp, save_embedding_layers=False)
    size = sum(os.path.getsize(os.path.join(tmp, f)) for f in os.listdir(tmp))
print(f"\nsaved adapter: {size / 2**20:.1f} MiB on disk; "
      f"the base model's weights: {sum(p.numel() for p in base_model.parameters()) * 2 / 2**30:.2f} GiB in 16-bit")

step  0  loss = 3.740


step  1  loss = 3.515


step  2  loss = 3.297


step  3  loss = 3.073


step  4  loss = 2.866


step  5  loss = 2.654


step  6  loss = 2.475


step  7  loss = 2.286


step  8  loss = 2.106


step  9  loss = 1.951

saved adapter: 2.1 MiB on disk; the base model's weights: 0.92 GiB in 16-bit


**Key takeaways:** the loss falls while more than 99.8% of the model stays frozen;
the saved artifact is an adapter of about 2 MiB, not a copy of the model; and one
base model can serve a different adapter per task. QLoRA applies the same idea to a
base model quantized to 4 bits, which is how 7B models are fine-tuned on a single
free Colab GPU.

## 4. Verifying AI-generated output: a checklist

LLM output is fluent whether or not it is correct. Before you trust (or submit)
anything a model produced, walk this list. It mirrors the course AI-use policy:
*you* are accountable for everything you turn in.

- [ ] **Can I explain every line?** If not, I don't submit it.
- [ ] **Did I run it?** On a case where I already know the correct answer, plus at least
  one edge case (empty input, wrong shape, boundary value).
- [ ] **Do the APIs actually exist?** Models invent plausible function names and
  signatures: check the real documentation.
- [ ] **Are factual claims grounded?** For RAG systems: is every statement in the answer
  supported by a retrieved chunk, and does each citation point at the right one?
- [ ] **Is the format valid?** Structured output (JSON, code) gets parsed and validated
  programmatically, never eyeballed.
- [ ] **Would it survive a re-run?** Re-ask the same question; confident answers that
  change on every run deserve suspicion.
- [ ] **Did I build an evaluation set?** Even 20 examples, re-run on every prompt or
  model change, beats impressions. Twenty to fifty is a smoke test, not a comparison.

**Further reading:** the Session 28 reading (*LLMs in Practice: Prompting, RAG, and
LoRA*) covers RAG against fine-tuning, LoRA's arithmetic, and evaluating LLM output
in more depth.